File: apps/mobile/src/ui/transcript/__tests__/pinReducer.test.ts
/**
 * Exhaustive pinReducer tests — bug C's cure is this state machine, so it gets
 * property-style storms on top of the case-by-case suite. The core invariant
 * throughout: a scroll may only ever be requested while `pinned`, and nothing
 * but a user reaching/leaving the bottom (or a session switch) flips `pinned`.
 */
import { describe, it, expect } from 'vitest';
import {
  initialPinState,
  pinReducer,
  type PinEvent,
  type PinState,
} from '../pinReducer';

const run = (events: PinEvent[], from: PinState = initialPinState): PinState =>
  events.reduce(pinReducer, from);

describe('pinReducer — basics', () => {
  it('starts pinned at bottom', () => {
    expect(initialPinState).toMatchObject({ pinned: true, atBottom: true, missedEntries: 0 });
  });

  it('unflagged scroll away from bottom unpins', () => {
    const s = run([{ type: 'user-scroll', atBottom: false }]);
    expect(s.pinned).toBe(false);
    expect(s.atBottom).toBe(false);
  });

  it('unflagged scroll landing on the bottom re-pins and clears missed count', () => {
    const s = run([
      { type: 'user-scroll', atBottom: false },
      { type: 'new-entries', count: 3 },
      { type: 'user-scroll', atBottom: true },
    ]);
    expect(s.pinned).toBe(true);
    expect(s.missedEntries).toBe(0);
  });

  it('reached-bottom pins regardless of how the viewport got there', () => {
    const s = run([
      { type: 'user-scroll', atBottom: false },
      { type: 'new-entries', count: 5 },
      { type: 'reached-bottom' },
    ]);
    expect(s.pinned).toBe(true);
    expect(s.atBottom).toBe(true);
    expect(s.missedEntries).toBe(0);
  });

  it('left-bottom alone never unpins (content growth is not user intent)', () => {
    const s = run([{ type: 'left-bottom' }]);
    expect(s.pinned).toBe(true);
    expect(s.atBottom).toBe(false);
  });

  it('session-switch resets to pinned but NOT at bottom from any state (CDX-024)', () => {
    const s = run([
      { type: 'user-scroll', atBottom: false },
      { type: 'new-entries', count: 9 },
      { type: 'programmatic-scroll-start' },
      { type: 'session-switch' },
    ]);
    // atBottom must be false: the owner effect scrolls iff pinned && !atBottom,
    // so a reset asserting atBottom would leave a long session mid-history with
    // no jump pill — the initial scrollToBottom has to fire.
    expect(s).toEqual({ ...initialPinState, atBottom: false });
  });
});

describe('pinReducer — programmatic-scroll flagging', () => {
  it('flagged scrolls never unpin', () => {
    const s = run([
      { type: 'programmatic-scroll-start' },
      { type: 'user-scroll', atBottom: false }, // mid-flight: not at bottom yet
      { type: 'user-scroll', atBottom: true },
      { type: 'programmatic-scroll-end' },
    ]);
    expect(s.pinned).toBe(true);
  });

  it('an unflagged scroll after the window closes unpins normally', () => {
    const s = run([
      { type: 'programmatic-scroll-start' },
      { type: 'user-scroll', atBottom: true },
      { type: 'programmatic-scroll-end' },
      { type: 'user-scroll', atBottom: false },
    ]);
    expect(s.pinned).toBe(false);
  });

  it('nested flags: still flagged until every start is ended', () => {
    const mid = run([
      { type: 'programmatic-scroll-start' },
      { type: 'programmatic-scroll-start' },
      { type: 'programmatic-scroll-end' },
    ]);
    expect(mid.programmaticDepth).toBe(1);
    const s = pinReducer(mid, { type: 'user-scroll', atBottom: false });
    expect(s.pinned).toBe(true); // still flagged
  });

  it('programmatic depth clamps at 0 on stray ends (double-end is a no-op)', () => {
    const s = run([
      { type: 'programmatic-scroll-end' },
      { type: 'programmatic-scroll-end' },
    ]);
    expect(s.programmaticDepth).toBe(0);
    // and a user scroll now unpins — the stray ends did not corrupt anything
    expect(pinReducer(s, { type: 'user-scroll', atBottom: false }).pinned).toBe(false);
  });

  it('flagged scroll to bottom while pinned keeps missed count at 0', () => {
    const s = run([
      { type: 'programmatic-scroll-start' },
      { type: 'user-scroll', atBottom: true },
      { type: 'programmatic-scroll-end' },
    ]);
    expect(s.missedEntries).toBe(0);
    expect(s.atBottom).toBe(true);
  });
});

describe('pinReducer — new entries', () => {
  it('never changes pinned', () => {
    expect(run([{ type: 'new-entries', count: 4 }]).pinned).toBe(true);
    const unpinned = run([
      { type: 'user-scroll', atBottom: false },
      { type: 'new-entries', count: 4 },
    ]);
    expect(unpinned.pinned).toBe(false);
  });

  it('counts missed entries only while unpinned', () => {
    const pinnedState = run([{ type: 'new-entries', count: 4 }]);
    expect(pinnedState.missedEntries).toBe(0);

    const s = run([
      { type: 'user-scroll', atBottom: false },
      { type: 'new-entries', count: 4 },
      { type: 'new-entries', count: 2 },
    ]);
    expect(s.missedEntries).toBe(6);
  });

  it('negative counts are ignored', () => {
    const s = run([
      { type: 'user-scroll', atBottom: false },
      { type: 'new-entries', count: -5 },
    ]);
    expect(s.missedEntries).toBe(0);
  });
});

// --- Property-style storms ---

/** Deterministic PRNG (mulberry32) so failures are reproducible by seed. */
function rng(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a |= 0; a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function randomEvent(rand: () => number): PinEvent {
  const roll = rand();
  if (roll < 0.35) return { type: 'user-scroll', atBottom: rand() < 0.4 };
  if (roll < 0.5) return { type: 'new-entries', count: 1 + Math.floor(rand() * 5) };
  if (roll < 0.62) return { type: 'programmatic-scroll-start' };
  if (roll < 0.74) return { type: 'programmatic-scroll-end' };
  if (roll < 0.84) return { type: 'reached-bottom' };
  if (roll < 0.94) return { type: 'left-bottom' };
  return { type: 'session-switch' };
}

describe('pinReducer — event storms (property-style)', () => {
  // 60k reducer steps: <1s alone, but worker contention in the full parallel
  // run can starve it past vitest's 5s default — give it explicit headroom.
  it('holds every invariant across 200 seeded random storms of 300 events', { timeout: 30_000 }, () => {
    for (let seed = 1; seed <= 200; seed++) {
      const rand = rng(seed);
      let state = initialPinState;
      // Model of the ONE owner effect: it "scrolls" only when pinned. We track
      // whether a scroll could ever be requested while unpinned — it must not.
      for (let i = 0; i < 300; i++) {
        const event = randomEvent(rand);
        const before = state;
        state = pinReducer(state, event);
        const ctx = `seed ${seed} step ${i} event ${event.type}`;

        // Invariant: depth never negative.
        expect(state.programmaticDepth, ctx).toBeGreaterThanOrEqual(0);
        // Invariant: missed entries never negative.
        expect(state.missedEntries, ctx).toBeGreaterThanOrEqual(0);
        // Invariant 1: an unflagged user scroll away from bottom always unpins.
        if (event.type === 'user-scroll' && !event.atBottom && before.programmaticDepth === 0) {
          expect(state.pinned, ctx).toBe(false);
        }
        // Invariant 2: reaching the bottom always pins + clears missed.
        if (event.type === 'reached-bottom'
            || (event.type === 'user-scroll' && event.atBottom && before.programmaticDepth === 0)) {
          expect(state.pinned, ctx).toBe(true);
          expect(state.missedEntries, ctx).toBe(0);
        }
        // Invariant 3: session-switch fully resets — pinned, but atBottom
        // false so the owner effect's scrollToBottom fires (CDX-024).
        if (event.type === 'session-switch') {
          expect(state, ctx).toEqual({ ...initialPinState, atBottom: false });
        }
        // Invariant 4: new-entries never flips pinned; while unpinned it never
        // "requests a scroll" — encoded as: unpinned stays unpinned and only
        // missedEntries may change.
        if (event.type === 'new-entries') {
          expect(state.pinned, ctx).toBe(before.pinned);
          if (!before.pinned) {
            expect({ ...state, missedEntries: 0 }, ctx).toEqual({ ...before, missedEntries: 0 });
          }
        }
        // Invariant 5: flagged scrolls never unpin.
        if (event.type === 'user-scroll' && before.programmaticDepth > 0) {
          expect(state.pinned, ctx).toBe(before.pinned);
        }
        // Global: pinned may only become false via an unflagged user scroll.
        if (!before.pinned || state.pinned || event.type === 'session-switch') continue;
        expect(event.type, ctx).toBe('user-scroll');
        expect(before.programmaticDepth, ctx).toBe(0);
      }
    }
  });

  it('interleaved programmatic windows and user flings: pin state converges to the geometry', () => {
    for (let seed = 300; seed < 340; seed++) {
      const rand = rng(seed);
      let state = initialPinState;
      // storm ends with: user scrolls to bottom → must be pinned
      for (let i = 0; i < 100; i++) state = pinReducer(state, randomEvent(rand));
      while (state.programmaticDepth > 0) state = pinReducer(state, { type: 'programmatic-scroll-end' });
      state = pinReducer(state, { type: 'user-scroll', atBottom: true });
      expect(state.pinned).toBe(true);
      expect(state.missedEntries).toBe(0);
      // ...and a storm ending with a scroll-away must be unpinned
      state = pinReducer(state, { type: 'user-scroll', atBottom: false });
      expect(state.pinned).toBe(false);
    }
  });
});

File: apps/android/app/src/test/java/com/codedeck/plus/ui/transcript/PinReducerTest.kt
package com.codedeck.plus.ui.transcript

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Exhaustive `pinReducer` tests — direct port of
 * `apps/mobile/src/ui/transcript/__tests__/pinReducer.test.ts`, including
 * the property-style storms (same mulberry32 PRNG, same seeds, same
 * invariants) so a failure here is reproducible against the TS suite's own
 * failure by seed number.
 */
class PinReducerTest {

    private fun run(events: List<PinEvent>, from: PinState = initialPinState): PinState =
        events.fold(from, ::pinReducer)

    // --- basics ---------------------------------------------------------

    @Test
    fun starts_pinned_at_bottom() {
        assertTrue(initialPinState.pinned)
        assertTrue(initialPinState.atBottom)
        assertEquals(0, initialPinState.missedEntries)
    }

    @Test
    fun unflagged_scroll_away_from_bottom_unpins() {
        val s = run(listOf(PinEvent.UserScroll(atBottom = false)))
        assertEquals(false, s.pinned)
        assertEquals(false, s.atBottom)
    }

    @Test
    fun unflagged_scroll_landing_on_the_bottom_re_pins_and_clears_missed_count() {
        val s = run(
            listOf(
                PinEvent.UserScroll(atBottom = false),
                PinEvent.NewEntries(3),
                PinEvent.UserScroll(atBottom = true),
            ),
        )
        assertTrue(s.pinned)
        assertEquals(0, s.missedEntries)
    }

    @Test
    fun reached_bottom_pins_regardless_of_how_the_viewport_got_there() {
        val s = run(
            listOf(
                PinEvent.UserScroll(atBottom = false),
                PinEvent.NewEntries(5),
                PinEvent.ReachedBottom,
            ),
        )
        assertTrue(s.pinned)
        assertTrue(s.atBottom)
        assertEquals(0, s.missedEntries)
    }

    @Test
    fun left_bottom_alone_never_unpins() {
        val s = run(listOf(PinEvent.LeftBottom))
        assertTrue(s.pinned)
        assertEquals(false, s.atBottom)
    }

    @Test
    fun session_switch_resets_to_pinned_but_not_at_bottom_from_any_state_cdx_024() {
        val s = run(
            listOf(
                PinEvent.UserScroll(atBottom = false),
                PinEvent.NewEntries(9),
                PinEvent.ProgrammaticScrollStart,
                PinEvent.SessionSwitch,
            ),
        )
        assertEquals(initialPinState.copy(atBottom = false), s)
    }

    // --- programmatic-scroll flagging ------------------------------------

    @Test
    fun flagged_scrolls_never_unpin() {
        val s = run(
            listOf(
                PinEvent.ProgrammaticScrollStart,
                PinEvent.UserScroll(atBottom = false),
                PinEvent.UserScroll(atBottom = true),
                PinEvent.ProgrammaticScrollEnd,
            ),
        )
        assertTrue(s.pinned)
    }

    @Test
    fun an_unflagged_scroll_after_the_window_closes_unpins_normally() {
        val s = run(
            listOf(
                PinEvent.ProgrammaticScrollStart,
                PinEvent.UserScroll(atBottom = true),
                PinEvent.ProgrammaticScrollEnd,
                PinEvent.UserScroll(atBottom = false),
            ),
        )
        assertEquals(false, s.pinned)
    }

    @Test
    fun nested_flags_still_flagged_until_every_start_is_ended() {
        val mid = run(
            listOf(
                PinEvent.ProgrammaticScrollStart,
                PinEvent.ProgrammaticScrollStart,
                PinEvent.ProgrammaticScrollEnd,
            ),
        )
        assertEquals(1, mid.programmaticDepth)
        val s = pinReducer(mid, PinEvent.UserScroll(atBottom = false))
        assertTrue(s.pinned) // still flagged
    }

    @Test
    fun programmatic_depth_clamps_at_0_on_stray_ends() {
        val s = run(listOf(PinEvent.ProgrammaticScrollEnd, PinEvent.ProgrammaticScrollEnd))
        assertEquals(0, s.programmaticDepth)
        assertEquals(false, pinReducer(s, PinEvent.UserScroll(atBottom = false)).pinned)
    }

    @Test
    fun flagged_scroll_to_bottom_while_pinned_keeps_missed_count_at_0() {
        val s = run(
            listOf(
                PinEvent.ProgrammaticScrollStart,
                PinEvent.UserScroll(atBottom = true),
                PinEvent.ProgrammaticScrollEnd,
            ),
        )
        assertEquals(0, s.missedEntries)
        assertTrue(s.atBottom)
    }

    // --- new entries ------------------------------------------------------

    @Test
    fun new_entries_never_changes_pinned() {
        assertTrue(run(listOf(PinEvent.NewEntries(4))).pinned)
        val unpinned = run(listOf(PinEvent.UserScroll(atBottom = false), PinEvent.NewEntries(4)))
        assertEquals(false, unpinned.pinned)
    }

    @Test
    fun new_entries_counts_missed_only_while_unpinned() {
        val pinnedState = run(listOf(PinEvent.NewEntries(4)))
        assertEquals(0, pinnedState.missedEntries)

        val s = run(
            listOf(
                PinEvent.UserScroll(atBottom = false),
                PinEvent.NewEntries(4),
                PinEvent.NewEntries(2),
            ),
        )
        assertEquals(6, s.missedEntries)
    }

    @Test
    fun new_entries_negative_counts_are_ignored() {
        val s = run(listOf(PinEvent.UserScroll(atBottom = false), PinEvent.NewEntries(-5)))
        assertEquals(0, s.missedEntries)
    }

    // --- property-style storms --------------------------------------------

    /** `Math.imul`-equivalent: Kotlin `Int * Int` already wraps to the low 32
     *  bits the same way, so this is only here to name the operation. */
    private fun imul(a: Int, b: Int): Int = a * b

    /** Deterministic PRNG (mulberry32) — same algorithm and seeds as
     *  `pinReducer.test.ts`'s own `rng()`, so a failure here reproduces
     *  against the TS suite's failure by seed number. Kotlin `Int` wraps on
     *  overflow the same way JS's `| 0` truncation does for this algorithm. */
    private fun mulberry32(seed: Int): () -> Double {
        var a = seed
        return {
            a += 0x6d2b79f5
            var t = a
            t = imul(t xor (t ushr 15), 1 or t)
            t = (t + imul(t xor (t ushr 7), 61 or t)) xor t
            ((t xor (t ushr 14)).toLong() and 0xFFFFFFFFL).toDouble() / 4294967296.0
        }
    }

    /** Cross-checks the Kotlin mulberry32 port against Node's own JS
     *  reference sequence (computed once, offline, from the exact algorithm
     *  `pinReducer.test.ts`'s `rng()` uses) — the storm tests below are only
     *  as reproducible-by-seed against the TS suite as this PRNG is faithful. */
    @Test
    fun mulberry32_matches_the_javascript_reference_sequence() {
        val seed1 = mulberry32(1)
        assertEquals(0.6270739405881613, seed1(), 1e-15)
        assertEquals(0.002735721180215478, seed1(), 1e-15)
        assertEquals(0.5274470399599522, seed1(), 1e-15)
        assertEquals(0.9810509674716741, seed1(), 1e-15)
        assertEquals(0.9683778982143849, seed1(), 1e-15)

        val seed2 = mulberry32(2)
        assertEquals(0.7342509443406016, seed2(), 1e-15)
        assertEquals(0.32499843230471015, seed2(), 1e-15)
        assertEquals(0.28529605525545776, seed2(), 1e-15)
        assertEquals(0.5379551574587822, seed2(), 1e-15)
        assertEquals(0.8752879470121115, seed2(), 1e-15)
    }

    private fun randomEvent(rand: () -> Double): PinEvent {
        val roll = rand()
        return when {
            roll < 0.35 -> PinEvent.UserScroll(atBottom = rand() < 0.4)
            roll < 0.5 -> PinEvent.NewEntries(1 + (rand() * 5).toInt())
            roll < 0.62 -> PinEvent.ProgrammaticScrollStart
            roll < 0.74 -> PinEvent.ProgrammaticScrollEnd
            roll < 0.84 -> PinEvent.ReachedBottom
            roll < 0.94 -> PinEvent.LeftBottom
            else -> PinEvent.SessionSwitch
        }
    }

    @Test
    fun holds_every_invariant_across_200_seeded_random_storms_of_300_events() {
        for (seed in 1..200) {
            val rand = mulberry32(seed)
            var state = initialPinState
            for (i in 0 until 300) {
                val event = randomEvent(rand)
                val before = state
                state = pinReducer(state, event)
                val ctx = "seed $seed step $i event $event"

                assertTrue(ctx, state.programmaticDepth >= 0)
                assertTrue(ctx, state.missedEntries >= 0)

                // Invariant 1: an unflagged user scroll away from bottom always unpins.
                if (event is PinEvent.UserScroll && !event.atBottom && before.programmaticDepth == 0) {
                    assertEquals(ctx, false, state.pinned)
                }
                // Invariant 2: reaching the bottom always pins + clears missed.
                val reachedBottom = event is PinEvent.ReachedBottom ||
                    (event is PinEvent.UserScroll && event.atBottom && before.programmaticDepth == 0)
                if (reachedBottom) {
                    assertEquals(ctx, true, state.pinned)
                    assertEquals(ctx, 0, state.missedEntries)
                }
                // Invariant 3: session-switch fully resets.
                if (event is PinEvent.SessionSwitch) {
                    assertEquals(ctx, initialPinState.copy(atBottom = false), state)
                }
                // Invariant 4: new-entries never flips pinned; while unpinned it
                // only ever changes missedEntries.
                if (event is PinEvent.NewEntries) {
                    assertEquals(ctx, before.pinned, state.pinned)
                    if (!before.pinned) {
                        assertEquals(ctx, before.copy(missedEntries = 0), state.copy(missedEntries = 0))
                    }
                }
                // Invariant 5: flagged scrolls never unpin.
                if (event is PinEvent.UserScroll && before.programmaticDepth > 0) {
                    assertEquals(ctx, before.pinned, state.pinned)
                }
                // Global: pinned may only become false via an unflagged user scroll.
                if (!before.pinned || state.pinned || event is PinEvent.SessionSwitch) continue
                assertTrue(ctx, event is PinEvent.UserScroll)
                assertEquals(ctx, 0, before.programmaticDepth)
            }
        }
    }

    @Test
    fun interleaved_programmatic_windows_and_user_flings_converge_to_the_geometry() {
        for (seed in 300 until 340) {
            val rand = mulberry32(seed)
            var state = initialPinState
            repeat(100) { state = pinReducer(state, randomEvent(rand)) }
            while (state.programmaticDepth > 0) state = pinReducer(state, PinEvent.ProgrammaticScrollEnd)
            state = pinReducer(state, PinEvent.UserScroll(atBottom = true))
            assertTrue(state.pinned)
            assertEquals(0, state.missedEntries)
            state = pinReducer(state, PinEvent.UserScroll(atBottom = false))
            assertEquals(false, state.pinned)
        }
    }
}
